When appending to a table, load if the dataframe contains a subset of the existing schema

mr-mcox opened the pull request:
## Purpose
Current behavior of `to_gbq` is fail if the schema of the new data is not equivalent to the current schema. However, this means that the load fails if the new data is missing columns that are present in the current schema. For instance, this may occur when the data source I am using to construct the dataframe only provides non-empty values. Rather than determining the current schema of the GBQ table and adding empty columns to my dataframe, I would like `to_gbq` to load my data if the columns in the dataframe are a subset of the current schema.

## Primary changes made

- Factoring a `schema` function out of `verify_schema` to support both `verify_schema` and  `schema_is_subset`
- `schema_is_subset` determines whether `local_schema` is a subset of `remote_schema`
- the append flag uses `schema_is_subset` rather than `verify_schema` to determine if the data can be loaded

## Auxiliary changes made

- `PROJECT_ID` etc are retrieved from an environment variable to facilitate local testing
- Running `test_gbq` through autopep8 added a row after two class names

MaximilianR commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -19,9 +19,9 @@
-PROJECT_ID = None
-PRIVATE_KEY_JSON_PATH = None
-PRIVATE_KEY_JSON_CONTENTS = None
+PROJECT_ID = os.getenv('PROJECT_ID')
```
I think the Google library pulls from `GOOGLE_CLOUD_PROJECT` (more confident about the newer library, which this package doesn't use yet). So you could either replicate that variable, or pass `None` through and the Google library should pick it up

MaximilianR commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -19,9 +19,9 @@
-PRIVATE_KEY_JSON_PATH = None
-PRIVATE_KEY_JSON_CONTENTS = None
+PROJECT_ID = os.getenv('PROJECT_ID')
+PRIVATE_KEY_JSON_PATH = os.getenv('PRIVATE_KEY_JSON_PATH')
```
Same as above for `GOOGLE_APPLICATION_CREDENTIALS`

MaximilianR commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -19,9 +19,9 @@
-PRIVATE_KEY_JSON_CONTENTS = None
+PROJECT_ID = os.getenv('PROJECT_ID')
+PRIVATE_KEY_JSON_PATH = os.getenv('PRIVATE_KEY_JSON_PATH')
+PRIVATE_KEY_JSON_CONTENTS = os.getenv('PRIVATE_KEY_JSON_CONTENTS')
```
(Contents doesn't have a Google version, so we do need to choose our own for that)

MaximilianR commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -1080,6 +1082,29 @@ def test_upload_data_if_table_exists_append(self):
+                   self.destination_table + test_id, _get_project_id(),
+                   if_exists='append', private_key=_get_private_key_path())
+
+        sleep(30)  # <- Curses Google!!!
```
We have a function `wait_for_job` which we copied from the Google docs; I'll put PR it in down the road (it relies on `google.cloud.bigquery`, so need to move some other code onto that)

jreback replied:
pls update as @MaximilianR suggests (look at other tests to see how this is done)

jreback replied:
@MaximilianR hmm, ``wait_for_job`` must be old.....

jreback commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -19,9 +19,9 @@
-PROJECT_ID = None
-PRIVATE_KEY_JSON_PATH = None
-PRIVATE_KEY_JSON_CONTENTS = None
+PROJECT_ID = os.getenv('GOOGLE_CLOUD_PROJECT')
```
https://pandas-gbq.readthedocs.io/en/latest/contributing.html are the docs for how to test this

parthea replied:
I am ok with these environment variables (They are just for local testing). I agree with @jreback that is also necessary to follow the steps in the contributing docs to configure Travis-CI to run on your pandas-gbq fork. You only need to configure Travis-CI once. After that unit tests run automatically for your pandas-gbq fork after each push.

jreback commented:
is there any reason to change ``if_exists``? IOW, is there a reason *not* to append with a compat schema? (albeit one that is larger), ever?

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -558,7 +558,9 @@ def load_data(self, dataframe, dataset_id, table_id, chunksize):
 
-    def verify_schema(self, dataset_id, table_id, schema):
+    def schema(self, dataset_id, table_id):
+        """Retrieve the schema of the table"""
```
can you add a doc-string

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -574,15 +576,26 @@ def verify_schema(self, dataset_id, table_id, schema):
             self.process_http_error(ex)
 
+    def verify_schema(self, dataset_id, table_id, schema):
+        fields_remote = set([json.dumps(field)
```
same here

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -574,15 +576,26 @@ def verify_schema(self, dataset_id, table_id, schema):
+        return fields_remote == fields_local
+
+    def schema_is_subset(self, dataset_id, table_id, schema):
+        fields_remote = set([json.dumps(field)
```
and here

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -574,15 +576,26 @@ def verify_schema(self, dataset_id, table_id, schema):
+                             for field in self.schema(dataset_id, table_id)])
+        fields_local = set(json.dumps(field_local)
+                           for field_local in schema['fields'])
+
```
these routines should prob just take the result of ``self.schema(dataset_id, table_id)``, yes?

jreback commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -1080,6 +1082,29 @@ def test_upload_data_if_table_exists_append(self):
                        private_key=_get_private_key_path())
 
+    def test_upload_subset_columns_if_table_exists_append(self):
+        test_id = "16"
```
can you add a comment with the issue ref number (this PR number is ok)

jreback commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -1275,6 +1300,18 @@ def test_verify_schema_ignores_field_mode(self):
             'Expected schema to match')
 
+    def test_retrieve_schema(self):
+        test_id = "15"
```
comment of issue number

jreback commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -1275,6 +1300,18 @@ def test_verify_schema_ignores_field_mode(self):
+        actual = self.sut.schema(self.dataset_prefix + "1", TABLE_ID + test_id)
+        expected = test_schema['fields']
+        assert expected == actual, 'Expected schema used to create table'
+
```
can you add some tests that exercise schema_is_subset

parthea commented:
> is there any reason to change if_exists? IOW, is there a reason not to append with a compat schema? (albeit one that is larger), ever?

@jreback I apologize if I missed the context for this comment. The default behaviour of the `if_exists` parameter is `fail` if a table already exists. Do we want to modify the default value of the `if_exists` parameter?

jreback commented:
@parthea 

>> is there any reason to change if_exists? IOW, is there a reason not to append with a compat schema? (albeit one that is larger), ever?

> @jreback I apologize if I missed the context for this comment. The default behaviour of the `if_exists` parameter is `fail` if a table already exists. Do we want to modify the default value of the `if_exists` parameter? 

what I mean is should this be a warning? you *could* have a 4th (maybe default state), like ``append_ok_schema_changes``. is that worthile?

parthea commented:
> what I mean is should this be a warning? you could have a 4th (maybe default state), like append_ok_schema_changes. is that worthile?

My initial thought is that we don't need to add another `if_exists` value. We could always add another value at a later stage if the community needs it.   While we are on the topic of the default state for the `if_exists` parameter, I do agree that the default value could be changed from `'fail'`  to `'append'`. I don't have a strong opinion about it.  @mr-mcox, thoughts ?

mr-mcox commented:
@parthea Regarding changing the `if_exists` default, I vote for keeping the default to `'fail'`. My rationale is that in BigQuery, removing records once you added them can be especially tricky. For the novice user, I'd want them to make a conscious choice to choose behavior that would add to an existing table.

@jreback has a good point in making the behavior of `'append'` also add new fields (not just adding a subset). This has been an occasional wish of mine and I may create a PR for that sometime in the future. However, I suspect that it may require some experimentation and require several more design considerations and ought to be out of the scope for this PR.

jreback commented:
ping, can you rebase and update to comments.

mr-mcox commented:
@jreback Sorry - I didn't realize that you were waiting on me for something. Can you be a little more specific about what you'd like me to do? Are you asking for me to take the five commits on this PR and rebase them to one, is there a comment that I neglected to address, or something else entirely?

jreback commented:
well needs a rebase, yes squashing is nice too (though not super necessary at this stage)

mr-mcox commented:
@jreback Alright - I've done a `git rebase upstream/master`, fixed a number of issues that resulted from that and squashed those commits. What would you like me to do next?

parthea commented:
@mr-mcox Thanks for the rebase. I'm really sorry for the delay in reviewing this pull request. I'm hoping to complete reviews for open pull requests this weekend.

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -557,7 +557,19 @@ def load_data(self, dataframe, dataset_id, table_id, chunksize):
+        Obtain from BigQuery the field names and field types
+        for the table defined by the parameters
+
+        :param str dataset_id: Name of the BigQuery dataset for the table
```
pls use the numpy-doc style (see all other doc-strings)

jreback commented on docs/source/changelog.rst:
```diff
@@ -9,6 +9,10 @@ Changelog
 
 - All gbq errors will simply be subclasses of ``ValueError`` and no longer inherit from the deprecated ``PandasError``.
 
+0.1.5 / 2017-04-20
```
move to latest (0.1.7)

jreback replied:
the issue reference is misformatted, see the other entries

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -573,15 +585,53 @@ def verify_schema(self, dataset_id, table_id, schema):
+        the schema passed in and indicate whether all fields in the former
+        are present in the latter. Order is not considered.
+
+        :param str dataset_id: Name of the BigQuery dataset for the table
```
same

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -573,15 +585,53 @@ def verify_schema(self, dataset_id, table_id, schema):
+        the schema passed in and indicate whether a subset of the fields in
+        the former are present in the latter. Order is not considered.
+
+        :param str dataset_id: Name of the BigQuery dataset for the table
```
same

jreback commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -1071,6 +1071,30 @@ def test_upload_data_if_table_exists_append(self):
                        private_key=_get_private_key_path())
 
+    def test_upload_subset_columns_if_table_exists_append(self):
+        # For pull request #24
```
be descriptive

mr-mcox commented:
@jreback Sorry this took so long. I did another rebase, updated the doc strings, added some description to test cases and adjusted the change log per your requested changes. The only thing I didn't do was to change `test_gbq.py` around line 1100. If I'm following the conversation correctly, @MaximilianR 's suggestion was to use `wait_for_job` but first needed to PR that code in. I searched the repository for`wait_for_job` and didn't find this function. You mentioned looking at other tests - the test I modeled that one after `test_upload_data_if_table_exists_replace` still uses the `sleep(30)` workaround, so I'm wondering if I'm misunderstanding what you're looking for. Could you clarify? Is there anything else you need me to do at this time?

jreback commented on docs/source/changelog.rst:
````diff
@@ -6,13 +6,15 @@ Changelog
 
 - Resolve issue where the optional ``--noauth_local_webserver`` command line argument would not be propagated during the authentication process. (:issue:`35`)
 - Drop support for Python 3.4 (:issue:`40`)
+- When using ```to_gbq``` if ```if_exists``` is set to ```append```, dataframe needs to contain only a subset of the fields in the BigQuery schema. (:issue: `24`)
````
no space after ``:issue:`24` ``, say more like:

``.to_gbq(...., if_exists='append')`` ...the rest of what you were saying

jreback commented on docs/source/changelog.rst:
```diff
@@ -6,13 +6,15 @@ Changelog
 
 - All gbq errors will simply be subclasses of ``ValueError`` and no longer inherit from the deprecated ``PandasError``.
 
-0.1.4 / 2017-03-17
```
did you change this? there isn't a 1.5 release

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -557,7 +557,25 @@ def load_data(self, dataframe, dataset_id, table_id, chunksize):
         self._print("\n")
 
-    def verify_schema(self, dataset_id, table_id, schema):
+    def schema(self, dataset_id, table_id):
```
can you update the ``to_gbq`` doc-string as well (with expl of the new behavior of if_exists)

MaximilianR commented:
This is `wait_for_job`:

```

def wait_for_job(job):

    # from https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/bigquery/cloud-client/snippets.py

    while True:
        job.reload()  # Refreshes the state via a GET request.
        if job.state == 'DONE':
            if job.error_result:
                raise RuntimeError(job.errors)
            return
        logger.info("Waiting for {} to complete".format(job))
        time.sleep(1)
```

jreback commented:
@MaximilianR thanks!

@mr-mcox if you want to add that as a utility function ok with that. pls run tests according to instructions and show the branch where they are passing.

(also ok to do this in another PR)

mr-mcox commented:
@jreback Good ideas on the updates to the changelog. You are correct that I removed 0.1.5 - this PR was the only change under that release and since there isn't 0.1.5 in the changelog on the master branch, I figured that it never happened and removed it entirely. I added `wait_for_job` with some minor tweaks, replaced the sleep functions across tests and the tests now complete in much less time. Since Travis CI on this page shows the tests on this branch passing, I'm not sure if you're looking for something else. That build is https://travis-ci.org/pydata/pandas-gbq/builds/240203212

mr-mcox commented:
Wait a sec - another build shows that I messed up calling `wait_for_job` - I'll fix that now.

mr-mcox commented:
@jreback I'm reverting that `wait_for_job` as it requires a job passed in. The job isn't returned by `to_gbq`, so I'm at a loss as to how to retrieve the job without refactoring `to_gbq`. @MaximilianR - what was your intention in using that utility? As I'm staring at `to_gbq`, I'm wondering if it is already waiting for the job to complete, though I vaguely recall from my early testing that the test failed without the wait.

MaximilianR commented:
Right, that's a good point:

- The current implementation of `to_gbq` uses streaming. This is OK, but slower (max 100k rows / sec), and more expensive. 
- The current implementation should complete very soon after returning - the data sits in a streaming buffer, and BQ queries that buffer before returning a result
- We use `upload_from_file` internally after writing to a CSV, which returns an async job, and so isn't ready immediately (though overall is faster)
- I would love to PR our stuff in, I've just been completely jammed here, and likely will be for another month
- Until we reconfigure `to_gbq` to use jobs, I would vote to allow `sleep`

jreback commented:
@parthea if you'd have a look
